internal/core/metadata/operators/device/update.go

```diff
@@ -0,0 +1,172 @@
+ * the License.
+ *******************************************************************************/
+
+package device
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/edgexfoundry/edgex-go/internal/core/metadata/errors"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/db"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+)
+
+type DeviceUpdateExecutor interface {
+	Execute() (err error)
+}
+
+type updateDevice struct {
+	database DeviceUpdater
+	device   contract.Device
+	events   chan DeviceEvent
+	logger   logger.LoggingClient
+}
+
+func NewUpdateDevice(ch chan DeviceEvent, db DeviceUpdater, d contract.Device, logger logger.LoggingClient) DeviceUpdateExecutor {
+	return updateDevice{database: db, events: ch, device: d, logger: logger}
+}
+
+func (op updateDevice) Execute() (err error) {
+	var evt DeviceEvent
+
+	// Check if the device exists
+	// First try ID
+	var oldDevice contract.Device
+	oldDevice, err = op.database.GetDeviceById(op.device.Id)
+	if err != nil {
+		// Then try name
+		oldDevice, err = op.database.GetDeviceByName(op.device.Name)
+		if err != nil {
+			if err == db.ErrNotFound {
+				err = errors.NewErrItemNotFound(fmt.Sprintf("device not found: %s %s", op.device.Name, op.device.Id))
+			}
+			op.logger.Error(err.Error())
+			evt.Error = err
+			op.events <- evt
+			return
+		}
+	}
+
+	if err = updateDeviceFields(op.device, &oldDevice, op.database); err != nil {
+		op.logger.Error(err.Error())
+		evt.Error = err
+		op.events <- evt
+		return
+	}
+
+	if err = op.database.UpdateDevice(oldDevice); err != nil {
+		op.logger.Error(err.Error())
+		evt.Error = err
+		op.events <- evt
+		return
+	}
+
+	// Device updated successfully.  Publish event onto supplied channel.
+
+	evt.DeviceId = op.device.Id
+	evt.DeviceName = op.device.Name
+	evt.HttpMethod = http.MethodPost
+	evt.ServiceId = op.device.Service.Id
+
+	op.events <- evt
+
+	return nil
+}
+
+func updateDeviceFields(from contract.Device, to *contract.Device, dbClient DeviceUpdater) error {
+
+	if (from.Service.String() != contract.DeviceService{}.String()) {
+		// Check if the new service exists
+		// Try ID first
+		ds, err := dbClient.GetDeviceServiceById(from.Service.Id)
+		if err != nil {
+			// Then try name
+			ds, err = dbClient.GetDeviceServiceByName(from.Service.Name)
+			if err != nil {
+				return errors.NewErrItemNotFound("Device service not found for updated device")
+			}
+		}
+
+		to.Service = ds
+	}
+	if (from.Profile.String() != contract.DeviceProfile{}.String()) {
+		// Check if the new profile exists
+		// Try ID first
+		dp, err := dbClient.GetDeviceProfileById(from.Profile.Id)
+		if err != nil {
+			// Then try Name
+			dp, err = dbClient.GetDeviceProfileByName(from.Profile.Name)
+			if err != nil {
+				return errors.NewErrItemNotFound("Device profile not found for updated device")
+			}
+		}
```

Comment:
Also, following from the comment above and the one I linked to on Anthony's PR, all of these database checks should be moved into Execute() prior to this function being called. When we call `updateDeviceFields` we should be confident that duplication checks and so on have been done already so we can just set the new values. As I said on the other PR, eventually the Device model should handle this itself. For that to occur, we don't want to burden the Device model with the need to connect to a database or some other service.